private/buf/bufgen/config.go

```diff
@@ -163,36 +163,67 @@ func validateExternalConfigV1(externalConfig ExternalConfigV1, id string) error
 		return fmt.Errorf("%s: no plugins set", id)
 	}
 	for _, plugin := range externalConfig.Plugins {
-		if plugin.Name == "" && plugin.Remote == "" {
-			return fmt.Errorf("%s: plugin name or remote plugin name is required", id)
+		numPluginIdentifiers := 0
+		var pluginIdentifier string
+		for _, possibleIdentifier := range []string{plugin.Plugin, plugin.Name, plugin.Remote} {
+			if possibleIdentifier != "" {
+				numPluginIdentifiers++
+				// Doesn't matter if we reassign here - we only allow one to be set below
+				pluginIdentifier = possibleIdentifier
+			}
+		}
+		if numPluginIdentifiers == 0 {
+			return fmt.Errorf("%s: one of plugin, name or remote is required", id)
 		}
-		if plugin.Name != "" && plugin.Remote != "" {
-			return fmt.Errorf("%s: only one of plugin name and remote plugin name can be set", id)
+		if numPluginIdentifiers > 1 {
+			return fmt.Errorf("%s: only one of plugin, name, or remote can be set", id)
 		}
 		if plugin.Out == "" {
-			return fmt.Errorf("%s: plugin %s out is required", id, plugin.Name)
+			return fmt.Errorf("%s: plugin %s out is required", id, pluginIdentifier)
 		}
-		if plugin.Remote != "" {
-			if _, _, _, _, err := bufremoteplugin.ParsePluginVersionPath(plugin.Remote); err != nil {
+		if plugin.Plugin != "" {
+			if _, err := bufpluginref.PluginReferenceForString(pluginIdentifier); err == nil {
+				// plugin.Plugin is a remote plugin
+				if err := checkPathAndStrategyUnset(id, plugin, pluginIdentifier); err != nil {
+					return err
+				}
```

Comment:
We should avoid as much client-side validation as possible - we've found ourselves in a variety of issues due to things like this. We might require that the fully-qualified plugin reference is used (i.e. it contains a version+revision), but we should only enforce that on the server-side.

That way, if we ever support a way to resolve the latest version by omitting a revision entirely, old CLIs will "just work".

I would only validate the absolute minimum set of values on the client side - the remote, owner, and name. We could even omit the version so that we could open the opportunity to automatically resolve the latest version AND revision.

This same thinking applies elsewhere. In summary let's try to keep the client-side validation to an absolute minimum to open up opportunities to permit new behavior with just server-side changes.

Reply:
I've updated the implementation to relax client side validation to allow either a PluginIdentity or a PluginReference to be specified.